File: PYTHON/Arrays/Day_11.py
# Hard questions are started

# Pascal Rectangle II
def getRow(rowIndex):
    answer = []
    ans = 1
    answer.append(ans)
    n = rowIndex +1
    for i in range(1,n):
        ans *= n-i
        ans /= i
        answer.append(ans)
    return answer

def generate(numRows):
    answer = []
    for i in range(0,numRows):
        answer.append(getRow(i))
    return answer


def majorityElement( nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        
        cnt1 = 0
        el1 = float("-inf")
        cnt2 = 0
        el2 = float("-inf")
        for num in nums:
            if(cnt1 == 0 and num != el2):
                cnt1 = 1
                el1 = num
            elif(cnt2 == 0 and num != el1):
                cnt2 = 1
                el2 = num
            elif(el1 == num):
                cnt1+=1
            elif(el2 == num):
                cnt2+=1
            else:
                cnt1-=1
                cnt2-=1

        ans = []
        cnt1 = 0
        cnt2 = 0
        for num in nums:
            if(el1 == num): 
                cnt1+=1
            if(el2 == num): 
                cnt2+=1

        mini = len(nums)//3
        if(cnt1 > mini): ans.append(el1)
        if(cnt2 > mini): ans.append(el2)

        return ans 

arr = [1,1,1,2,2,3,3,3]
print(majorityElement(arr))

print(generate(4))
        
        

File: JAVA/Arrays/Day_11.java
import java.util.ArrayList;

public class Day_11{
    public static ArrayList<Integer> Majority_Element_II(int[] arr){
        int cnt1 = 0, cnt2 = 0;
        int el1 = Integer.MIN_VALUE, el2 = Integer.MIN_VALUE;
        for(int num : arr){
            if(cnt1 == 0 && el2 != num){
                cnt1 = 1;
                el1 = num;
            }
            else if(cnt2 == 0 && el1 != num){
                cnt2 = 1;
                el2 = num;
            }
            else if(el1 == num){
                cnt1++;
            }
            else if(el2 == num){
                cnt2++;
            }
            else{
                cnt1--;
                cnt2--;
            }
        }

        ArrayList<Integer> Answer = new ArrayList<>();
        cnt1 = 0;
        cnt2 = 0;
        int mini = arr.length/3;
        for(int num : arr){
            if(el1 == num) cnt1++;
            if(el2 == num) cnt2++;
        }
        if(cnt1 > mini) Answer.add(el1);
        if(cnt2 > mini) Answer.add(el2);
        return Answer;
    }
    public static void main(String[] args) {
        int[] arr = {1,1,1,2,2,3,3,3};
        System.out.println(Majority_Element_II(arr));
    }
}